pkg/openstack/loadbalancer_test.go

```diff
@@ -973,3 +973,175 @@ func TestLbaasV2_updateServiceAnnotations(t *testing.T) {
+			},
+			expectedLen: 2,
+			expectError: false,
+			testType:    validNodesHttpMonitorOff,
+		},
+		{
+			name:  "Valid nodes, canUseHTTPMonitor=true",
+			nodes: []*corev1.Node{node1, node2},
+			port:  corev1.ServicePort{NodePort: 8080},
+			svcConf: &serviceConfig{
+				preferredIPFamily:   corev1.IPv4Protocol,
+				lbMemberSubnetID:    "subnet-12345-test",
+				healthCheckNodePort: 8081,
+				enableMonitor:       true,
+			},
+			expectedLen: 2,
+			expectError: false,
+			testType:    validNodesHttpMonitorOn,
+		},
+		{
+			name:  "Invalid preferred IP family, fallback to default",
+			nodes: []*corev1.Node{node1, node2},
+			port:  port,
+			svcConf: &serviceConfig{
+				preferredIPFamily:   "invalid-family",
+				lbMemberSubnetID:    "subnet-12345-test",
+				healthCheckNodePort: 8081,
+			},
+			expectedLen: 2,
+			expectError: false,
+			testType:    invalidPreferredIPFamily,
+		},
+		{
+			name:  "ErrNoAddressFound happens and a member is ignored",
+			nodes: []*corev1.Node{},
+			port:  corev1.ServicePort{NodePort: 8080},
+			svcConf: &serviceConfig{
+				preferredIPFamily:   corev1.IPv4Protocol,
+				lbMemberSubnetID:    "subnet-12345-test",
+				healthCheckNodePort: 8081,
+				enableMonitor:       false,
+			},
+			expectedLen:          0,
+			expectError:          true,
+			checkNewMembersEmpty: true,
+			testType:             errNoAddressFoundMemberIgnored,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			switch tc.testType {
+			case nodePortEquals0:
+				members, _, err := lbaas.buildBatchUpdateMemberOpts(tc.port, tc.nodes, tc.svcConf)
+				assert.NoError(t, err, "Expected no error")
+				assert.Len(t, members, tc.expectedLen, "Unexpected number of members")
+			case errNoAddressFound:
+				members, _, err := lbaas.buildBatchUpdateMemberOpts(tc.port, tc.nodes, tc.svcConf)
+				assert.NoError(t, err, "Expected no error")
+				assert.Nil(t, members, "Unexpected number of members")
+			case validNodesHttpMonitorOff:
+				members, _, err := lbaas.buildBatchUpdateMemberOpts(tc.port, tc.nodes, tc.svcConf)
+				assert.NoError(t, err, "Expected no error")
+				assert.Len(t, members, tc.expectedLen, "Unexpected number of members")
+			case validNodesHttpMonitorOn:
+				members, _, err := lbaas.buildBatchUpdateMemberOpts(tc.port, tc.nodes, tc.svcConf)
+				assert.NoError(t, err, "Expected no error")
+				assert.Len(t, members, tc.expectedLen, "Unexpected number of members")
+			case invalidPreferredIPFamily:
+				members, _, err := lbaas.buildBatchUpdateMemberOpts(tc.port, tc.nodes, tc.svcConf)
+				assert.NoError(t, err, "Expected no error")
+				assert.Len(t, members, tc.expectedLen, "Unexpected number of members")
+			case errNoAddressFoundMemberIgnored:
+				members, _, err := lbaas.buildBatchUpdateMemberOpts(tc.port, tc.nodes, tc.svcConf)
+				assert.NoError(t, err, "Expected no error")
+				assert.Len(t, members, tc.expectedLen, "Unexpected number of members")
+			}
```

Comment:
In this PR, you are defining a slice of cases and then you `range` over it. Which means that you want to apply the same testing logic multiple times. But then you use a combination of `iota` constants and `switch` to define a separate behaviour for each test case. Which means that you haven't found that common testing logic.

If I read this correctly, each `case` here do exactly the same, except `case errNoAddressFound` where you `assert.Nil`. Are you sure that `assert.Len(t, members, 0)` doesn't do what you want? Because if that works, then you can do the same for all cases:

```Go
members, _, err := lbaas.buildBatchUpdateMemberOpts(tc.port, tc.nodes, tc.svcConf)
assert.NoError(t, err)
assert.Len(t, members, tc.expectedLen)
```

---

Also a nit: the error messages you pass as arguments to the assertions don't add any context specific to the case: you can remove them and let the library provide a generic error message.

Reply:
Fixed. Please review @pierreprinetti